### PR TITLE
chore(web): add image-based lighting

### PR DESCRIPTION
**Pull request by airslice**
# Overview

Supports image-based lighting for beta.

## What I've done

Add image based lighting properties to scene, which should effect PBR models & 3d tiles in scene:
```ts
type SceneProperty = {
  light?: {
    // ...
    specularEnvironmentMaps?: string, // URL to a KTX2 format
    sphericalHarmonicCoefficients?: [x: number, y: number, z: number][],
    imageBasedLightIntensity?: number, // Default: 1
  };
};
```

Add image based lighting properties to 3d tiles appearance & model appearance
```ts
type Cesium3DTilesAppearance = {
  // ...
  specularEnvironmentMaps?: string;
  sphericalHarmonicCoefficients?: [x: number, y: number, z: number][];
  imageBasedLightIntensity?: number;
}

type ModelAppearance = {
  // ...
  specularEnvironmentMaps?: string;
  sphericalHarmonicCoefficients?: [x: number, y: number, z: number][];
  imageBasedLightIntensity?: number;
}
```

### Scene IBL

```js
let layerId = reearth.layers.add({
  type: "simple",
  data: {
    type: "3dtiles",
    url: "https://assets.cms.plateau.reearth.io/assets/11/6d05db-ed47-4f88-b565-9eb385b1ebb0/13100_tokyo23-ku_2022_3dtiles%20_1_1_op_bldg_13101_chiyoda-ku_lod1/tileset.json",
  },
  "3dtiles": {
  }
});
setTimeout(()=>{
  reearth.camera.flyTo(layerId, {duration: 0});
},10);
```

```js
reearth.scene.overrideProperty({
    light: {
      sphericalHarmonicCoefficients: [
        [0.499745965003967, 0.499196201562881, 0.500154078006744],
        [0.265826553106308, -0.266099184751511, 0.265922993421555],
        [0.243236944079399, 0.266723394393921, -0.265380442142487],
        [-0.266895800828934, 0.265416264533997, 0.266921550035477],
        [0.000195000306121, -0.000644546060357, -0.000383183418307],
        [-0.000396036746679, -0.000622032093816, 0.000262127199676],
        [-0.000214280473301, 0.00004872302452, -0.000059724134189],
        [0.000107143961941, -0.000126510843984, -0.000425444566645],
        [-0.000069071611506, 0.000134039684781, -0.000119135256682],
      ],
      imageBasedLightIntensity: 1,
    },
});
```

```js
reearth.scene.overrideProperty({
    light: {
      specularEnvironmentMaps: "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx2"
    },
  });
```

```js
reearth.scene.overrideProperty({
    light: {
    },
});
```

### 3d tiles IBL

```js
let layerId2 = reearth.layers.add({
  type: "simple",
  data: {
    type: "3dtiles",
    url: "https://assets.cms.plateau.reearth.io/assets/5d/134c5b-d8a0-4cca-aeff-2354188a17ca/13100_tokyo23-ku_2022_3dtiles%20_1_1_op_bldg_13102_chuo-ku_lod1/tileset.json",
  },
  "3dtiles": {}
});
setTimeout(()=>{
  reearth.camera.flyTo(layerId2, {duration: 0});
},10);
```

```js
reearth.layers.override(layerId2, {
  "3dtiles": {
    sphericalHarmonicCoefficients: [
        [0.499745965003967, 0.499196201562881, 0.500154078006744],
        [0.265826553106308, -0.266099184751511, 0.265922993421555],
        [0.243236944079399, 0.266723394393921, -0.265380442142487],
        [-0.266895800828934, 0.265416264533997, 0.266921550035477],
        [0.000195000306121, -0.000644546060357, -0.000383183418307],
        [-0.000396036746679, -0.000622032093816, 0.000262127199676],
        [-0.000214280473301, 0.00004872302452, -0.000059724134189],
        [0.000107143961941, -0.000126510843984, -0.000425444566645],
        [-0.000069071611506, 0.000134039684781, -0.000119135256682],
      ],
      imageBasedLightIntensity: 1,
  },
});
```

```js
reearth.layers.override(layerId2, {
  "3dtiles": {
    sphericalHarmonicCoefficients: [],
  },
});
```

```js
reearth.layers.override(layerId2, {
  "3dtiles": {
    sphericalHarmonicCoefficients: [],
    specularEnvironmentMaps: "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx2"
  },
});
```

### Model IBL

```js
const modelLayerId = reearth.layers.add({
  type: "simple",
  data: {
    type: "geojson",
    value: {
      type: "Feature",
      geometry: {
        type: "Point",
        coordinates: [139.753214, 35.684891, 1], // Japan Tokyo Chiyoda-ku
      },
    },
  },
  model: {
    url: "https://api.test.reearth.dev/assets/01h5478z9t8r9cshbwvpxepwdd.glb",
    scale: 1000,
  },
});
setTimeout(()=>{
  reearth.camera.flyTo(modelLayerId, {duration: 0});
},10);
```

```js
reearth.layers.override(modelLayerId, {
  model: {
    sphericalHarmonicCoefficients: [
        [0.499745965003967, 0.499196201562881, 0.500154078006744],
        [0.265826553106308, -0.266099184751511, 0.265922993421555],
        [0.243236944079399, 0.266723394393921, -0.265380442142487],
        [-0.266895800828934, 0.265416264533997, 0.266921550035477],
        [0.000195000306121, -0.000644546060357, -0.000383183418307],
        [-0.000396036746679, -0.000622032093816, 0.000262127199676],
        [-0.000214280473301, 0.00004872302452, -0.000059724134189],
        [0.000107143961941, -0.000126510843984, -0.000425444566645],
        [-0.000069071611506, 0.000134039684781, -0.000119135256682],
      ],
    imageBasedLightIntensity: 0.5,
    specularEnvironmentMaps: '',
  },
});
```

```js
reearth.layers.override(modelLayerId, {
  model: {
    sphericalHarmonicCoefficients: '',
    specularEnvironmentMaps: "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx2",
  },
});
```

```js
reearth.layers.override(modelLayerId, {
  model: {
    sphericalHarmonicCoefficients: [],
    specularEnvironmentMaps: '',
  },
});
```

## Memo

- Our override rule seems different on scene property and layer property. 
 - Override scene property:
   - `undefined` can be used.
   - overridden property will replace the previous value.
 - Override layer property:
   - `undefined` will be ignored.
   - overridden property will be merged with the previous value.

### CODE REVIEW

File: web/src/beta/lib/core/engines/Cesium/Feature/index.tsx [hunk 1]
```diff
@@ -160,10 +160,14 @@ export default function Feature({
         {displayType.map(k => {
           const [C] = components[k] ?? [];
           const isVisible = layer.layer.visible !== false && !isHidden;
+          const useSphericalHormonicCoefficients =
+            !!props.sceneProperty.light?.sphericalHarmonicCoefficients;
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/index.tsx:164]:
```suggestion
            !!props.sceneProperty.light?.sphericalHarmonicCoefficients && !!props.sceneProperty.light?.specularEnvironmentMaps;
```
I think we should handle the environment maps in here.

File: web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts [hunk 2]
```diff
@@ -516,11 +529,34 @@ export const useHooks = ({
       : null;
   }, [isVisible, tileset, url, type, meta]);
 
+  const imageBasedLighting = useMemo(() => {
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts:532]: I think we should support this for model as well:pray:
Then we can test the environment map more easily by using Cesium sandcastle's glb file.
https://sandcastle.cesium.com/index.html?src=Image-Based%20Lighting.html

**Review by pyshx — APPROVED** [review]

File: web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts (hunk 2 above)
Comment (airslice) [reply web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts:532]: Thank you for helping implement the IBL for model!

File: web/src/beta/lib/core/engines/Cesium/Feature/Model/index.tsx [hunk 3]
```diff
@@ -94,6 +122,72 @@ export default function Model({ id, isVisible, property, geometry, layer, featur
...
+
+    const ibl = new ImageBasedLighting();
+    const specularEnvironmentMaps = property?.specularEnvironmentMaps
+      ? property?.specularEnvironmentMaps
+      : sceneProperty?.light?.specularEnvironmentMaps;
+    const sphericalHarmonicCoefficients = property?.sphericalHarmonicCoefficients
+      ? property?.sphericalHarmonicCoefficients
+      : sceneProperty?.light?.sphericalHarmonicCoefficients;
+    const imageBasedLightIntensity =
+      property?.imageBasedLightIntensity ?? sceneProperty?.light?.imageBasedLightIntensity;
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/Model/index.tsx:142]:
```suggestion
    const specularEnvironmentMaps =
      property?.specularEnvironmentMaps ?? sceneProperty?.light?.specularEnvironmentMaps;
    const sphericalHarmonicCoefficients =
      property?.sphericalHarmonicCoefficients ?? sceneProperty?.light?.sphericalHarmonicCoefficients;
    const imageBasedLightIntensity =
      property?.imageBasedLightIntensity ?? sceneProperty?.light?.imageBasedLightIntensity;
```
Maybe we can align the expression of the code?

File: web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts [hunk 4]
```diff
@@ -516,11 +530,53 @@ export const useHooks = ({
...
+
+    const ibl = new ImageBasedLighting();
+    const specularEnvironmentMaps = property?.specularEnvironmentMaps
+      ? property?.specularEnvironmentMaps
+      : sceneProperty?.light?.specularEnvironmentMaps;
+    const sphericalHarmonicCoefficients = property?.sphericalHarmonicCoefficients
+      ? property?.sphericalHarmonicCoefficients
+      : sceneProperty?.light?.sphericalHarmonicCoefficients;
+    const imageBasedLightIntensity =
+      property?.imageBasedLightIntensity ?? sceneProperty?.light?.imageBasedLightIntensity;
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts:550]:
```suggestion
    const specularEnvironmentMaps =
      property?.specularEnvironmentMaps ?? sceneProperty?.light?.specularEnvironmentMaps;
    const sphericalHarmonicCoefficients =
      property?.sphericalHarmonicCoefficients ?? sceneProperty?.light?.sphericalHarmonicCoefficients;
    const imageBasedLightIntensity =
      property?.imageBasedLightIntensity ?? sceneProperty?.light?.imageBasedLightIntensity;
```
Maybe we can align the expression of the code?

File: web/src/beta/lib/core/engines/Cesium/Feature/index.tsx [hunk 5]
```diff
@@ -161,9 +167,23 @@ export default function Feature({
...
 
+          const useSceneSphericalHarmonicCoefficients =
+            !!props.sceneProperty?.light?.sphericalHarmonicCoefficients;
+          const useSceneSpecularEnvironmentMaps =
+            !!props.sceneProperty?.light?.specularEnvironmentMaps;
+
+          const use3dtilesIBL =
+            layer?.layer?.type === "simple" &&
+            (!!layer?.layer?.["3dtiles"]?.sphericalHarmonicCoefficients ||
+              !!layer?.layer?.["3dtiles"]?.specularEnvironmentMaps);
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/index.tsx:178]:
```suggestion
          // NOTE: IBL for 3dtiles is not updated unless Tileset feature component is re-created. 
          const useSceneSphericalHarmonicCoefficients =
            !!props.sceneProperty?.light?.sphericalHarmonicCoefficients;
          const useSceneSpecularEnvironmentMaps =
            !!props.sceneProperty?.light?.specularEnvironmentMaps;
          const use3dtilesIBL =
            layer?.layer?.type === "simple" &&
            (!!layer?.layer?.["3dtiles"]?.sphericalHarmonicCoefficients ||
              !!layer?.layer?.["3dtiles"]?.specularEnvironmentMaps);
```

File: web/src/beta/lib/core/engines/Cesium/Feature/utils.tsx [hunk 6]
```diff
@@ -39,7 +45,7 @@ export type FeatureProps<P = any> = {
   layer?: ComputedLayer;
   feature?: ComputedFeature;
   geometry?: Geometry;
-  sceneProperty?: any;
+  sceneProperty?: SceneProperty;
```
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/utils.tsx:48]: 🥇

File: web/src/beta/lib/core/engines/Cesium/Feature/index.tsx (hunk 5 above)
Comment (keiya01) [inline web/src/beta/lib/core/engines/Cesium/Feature/index.tsx:178]:
```suggestion
          const use3dtilesSphericalHarmonicCoefficients =
            layer?.layer?.type === "simple" && !!layer?.layer?.["3dtiles"]?.sphericalHarmonicCoefficients;
         const use3dtilesSpecularEnvironmentMaps =
            layer?.layer?.type === "simple" && !!layer?.layer?.["3dtiles"]?.specularEnvironmentMaps;
```
I think we should separate the key to make sure to update the component.

Comment (keiya01) [reply web/src/beta/lib/core/engines/Cesium/Feature/index.tsx:178]: For example, `3d tiles IBL` seems not to be working.

File: web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts (hunk 4 above)
Comment (airslice) [reply web/src/beta/lib/core/engines/Cesium/Feature/Tileset/hooks.ts]: Updated. 🙏🏻

File: web/src/beta/lib/core/engines/Cesium/Feature/Model/index.tsx (hunk 3 above)
Comment (airslice) [reply web/src/beta/lib/core/engines/Cesium/Feature/Model/index.tsx]: Updated. 🙏🏻

File: web/src/beta/lib/core/engines/Cesium/Feature/index.tsx (hunk 5 above)
Comment (airslice) [reply web/src/beta/lib/core/engines/Cesium/Feature/index.tsx]: Updated. 🙏🏻

**Review by keiya01 — APPROVED** [review]

**Pull request merged**